### PR TITLE
[SPARK-9225] [MLlib] LDASuite needs unit tests for empty documents

### PR DESCRIPTION
**Pull request by rotationsymmetry**
Add unit tests for running LDA with empty documents.
Both EMLDAOptimizer and OnlineLDAOptimizer are tested.

@feynmanliang

### CODE REVIEW

**Comment by yu-iskw** [comment]
@rotationsymmetry Great work! We could also implement those two tests in one test suite with for-loop of the optimizers instead. Since both of them are almost same and there is a difference about the optimizers.

@feynmanliang What do you think?

**Comment by feynmanliang** [comment]
LGTM pending tests, we'll need @jkbradley to trigger the Jenkins build.

I would actually prefer that the two be in separate tests so it's easy to identify which optimizer is failing by just looking at which test is failing. I agree there is quite a bit of DRYing to do in `LDASuite`, but that can be addressed in some follow up PRs.

**Comment by yu-iskw** [comment]
@feynmanliang I got it. Thanks.

**Comment by rotationsymmetry** [comment]
@yu-iskw @feynmanliang Thanks for your review!

@jkbradley Will you please trigger the Jenkins build?

**Comment by yu-iskw** [comment]
Jenkins, test this please.

**Comment by mengxr** [comment]
ok to test

**Comment by rotationsymmetry** [comment]
@mengxr Thank you for OK to test.

I checked the Jerkins log. The failure is caused by git unable to fetch the repo. Anything on my end to fix it? 

cc @feynmanliang  @yu-iskw 

From Jenkins log:
```
GitHub pull request #7620 of commit ad55665f083025074bf3f350a1a66baab9d4a4be automatically merged.
[EnvInject] - Loading node environment variables.
Building remotely on amp-jenkins-worker-07 (centos) in workspace /home/jenkins/workspace/SparkPullRequestBuilder
 > git rev-parse --is-inside-work-tree # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/apache/spark.git # timeout=10
Fetching upstream changes from https://github.com/apache/spark.git
 > git --version # timeout=10
 > git fetch --tags --progress https://github.com/apache/spark.git +refs/pull/7620/*:refs/remotes/origin/pr/7620/* # timeout=15
ERROR: Timeout after 15 minutes
ERROR: Error fetching remote repo 'origin'
hudson.plugins.git.GitException: Failed to fetch from https://github.com/apache/spark.git
	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:735)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:983)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1016)
	at hudson.scm.SCM.checkout(SCM.java:485)
	at hudson.model.AbstractProject.checkout(AbstractProject.java:1282)
	at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:610)
	at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:532)
	at hudson.model.Run.execute(Run.java:1741)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:381)
Caused by: hudson.plugins.git.GitException: Command "git fetch --tags --progress https://github.com/apache/spark.git +refs/pull/7620/*:refs/remotes/origin/pr/7620/*" returned status code 143:
stdout: 
stderr: error: RPC failed; result=18, HTTP code = 200

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1591)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1379)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$300(CliGitAPIImpl.java:86)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:324)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:152)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:145)
	at hudson.remoting.UserRequest.perform(UserRequest.java:121)
	at hudson.remoting.UserRequest.perform(UserRequest.java:49)
	at hudson.remoting.Request$2.run(Request.java:325)
	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
	at ......remote call to amp-jenkins-worker-07(Native Method)
	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1361)
	at hudson.remoting.UserResponse.retrieve(UserRequest.java:221)
	at hudson.remoting.Channel.call(Channel.java:753)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:145)
	at sun.reflect.GeneratedMethodAccessor230.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:131)
	at com.sun.proxy.$Proxy54.execute(Unknown Source)
	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:733)
	... 11 more
ERROR: Error fetching remote repo 'origin'
Archiving artifacts
Recording test results
ERROR: Publisher 'Publish JUnit test result report' failed: Test reports were found but none of them are new. Did tests run? 
For example, /home/jenkins/workspace/SparkPullRequestBuilder/bagel/target/test-reports/org.apache.spark.bagel.BagelSuite.xml is 2 hr 19 min old

Finished: FAILURE
```

**Comment by mengxr** [comment]
test this please

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 1]
```diff
@@ -324,6 +324,58 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
     }
   }
 
+  test("Running EM LDA Optimizer with Empty Docs") {
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: minor: `test("EMLDAOptimizer with empty docs")`

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 2]
```diff
@@ -324,6 +324,58 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
...
+    val vocabSize = 6
+
+    def emptyDocs: Array[(Long, Vector)] = Array(
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]:
~~~
val emptyDocs = Seq.fill(6)(Vectors.sparse(vocabSize, Array.empty, Array.empty))
  .zipWithIndex { case (wordCounts, docId) =>
    (docId.toLong, wordCount)
  }
~~~

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 3]
```diff
@@ -324,6 +324,58 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
...
+      Vectors.sparse(vocabSize, Array(), Array()),
+      Vectors.sparse(vocabSize, Array(), Array()),
+      Vectors.sparse(vocabSize, Array(), Array()),
+      Vectors.sparse(vocabSize, Array(), Array()),
+      Vectors.sparse(vocabSize, Array(), Array()),
+      Vectors.sparse(vocabSize, Array(), Array())
+    ).zipWithIndex.map { case (wordCounts, docId) => (docId.toLong, wordCounts) }
+
+    val distributedEmptyDocs = sc.parallelize(emptyDocs)
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: `sc.parallelize(emptyDocs, 2)` to force more than one partitions

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 4]
```diff
@@ -324,6 +324,58 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
...
+
+    val op = new EMLDAOptimizer()
+
+    val lda = new LDA()
+    lda.setK(3)
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: remove `lda.` and call setters directly

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 5]
```diff
@@ -324,6 +324,58 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
...
+      .setMaxIterations(5)
+      .setSeed(12345)
+      .setOptimizer(op)
+
+    val model = lda.run(distributedEmptyDocs)
+    assert(model.vocabSize === vocabSize)
+  }
+
+  test("Running Online LDA Optimizer with Empty Docs") {
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: minor: `test("OnlineLDAOptimizer with empty docs")`

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 6]
```diff
@@ -324,6 +324,58 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
...
+    val vocabSize = 6
+
+    def emptyDocs: Array[(Long, Vector)] = Array(
+      Vectors.sparse(vocabSize, Array(), Array()),
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: ditto

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 7]
```diff
@@ -324,6 +324,58 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
...
+      Vectors.sparse(vocabSize, Array(), Array()),
+      Vectors.sparse(vocabSize, Array(), Array()),
+      Vectors.sparse(vocabSize, Array(), Array()),
+      Vectors.sparse(vocabSize, Array(), Array()),
+      Vectors.sparse(vocabSize, Array(), Array())
+    ).zipWithIndex.map { case (wordCounts, docId) => (docId.toLong, wordCounts) }
+
+    val distributedEmptyDocs = sc.parallelize(emptyDocs)
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: ditto

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 8]
```diff
@@ -324,6 +324,58 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
...
+
+    val op = new OnlineLDAOptimizer()
+
+    val lda = new LDA()
+    lda.setK(3)
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: ditto

**Comment by rotationsymmetry** [comment]
@mengxr Thanks a lot for your comments! It is very helpful. I have incorporated them in the commit.

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 9]
```diff
@@ -339,6 +339,52 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
     }
   }
 
+  test("EMLDAOptimizer with empty docs") {
+    val vocabSize = 6
+
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: minor: it is not necessary to have extra blank lines in Spark code base

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 10]
```diff
@@ -339,6 +339,52 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
...
+    val emptyDocsArray = Array.fill(6)(Vectors.sparse(vocabSize, Array.empty, Array.empty))
+    val emptyDocs: Array[(Long, Vector)] = emptyDocsArray
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: type info is not necessary

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 11]
```diff
@@ -339,6 +339,52 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
...
+      .zipWithIndex.map { case (wordCounts, docId) =>
+      (docId.toLong, wordCounts)
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: fix indentation (check my previous comment)

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!

**Pull request closed without merging**